editor/scene/2d/particles_2d_editor_plugin.cpp

```diff
@@ -61,154 +63,257 @@ void GPUParticles2DEditorPlugin::_add_menu_options(PopupMenu *p_menu) {
+	emission_mask_dialog = memnew(ConfirmationDialog);
+	emission_mask_dialog->set_title(TTRC("Load Emission Mask"));
+	emission_mask_dialog->add_child(file_dialog);
+	emission_mask_dialog->get_ok_button()->set_disabled(true);
 
 	VBoxContainer *emvb = memnew(VBoxContainer);
-	emission_mask->add_child(emvb);
+	emission_mask_dialog->add_child(emvb);
+
+	HBoxContainer *mask_img_hbox = memnew(HBoxContainer);
+
+	mask_img_path_line_edit = memnew(LineEdit);
+	mask_img_hbox->add_child(mask_img_path_line_edit);
+	mask_img_path_line_edit->set_h_size_flags(Control::SIZE_EXPAND_FILL);
+	mask_img_path_line_edit->set_editable(false);
+	mask_img_path_line_edit->connect(SceneStringName(text_changed), callable_mp(this, &Particles2DEditorPlugin::_validate_textures).unbind(1));
+
+	mask_browse_button = memnew(Button);
+	mask_img_hbox->add_child(mask_browse_button);
+	mask_browse_button->connect(SceneStringName(pressed), callable_mp(this, &Particles2DEditorPlugin::_browse_mask_texture_pressed));
+	emvb->add_margin_child(TTRC("Mask Texture"), mask_img_hbox);
 
 	emission_mask_mode = memnew(OptionButton);
-	emission_mask_mode->add_item(TTR("Solid Pixels"), EMISSION_MODE_SOLID);
-	emission_mask_mode->add_item(TTR("Border Pixels"), EMISSION_MODE_BORDER);
-	emission_mask_mode->add_item(TTR("Directed Border Pixels"), EMISSION_MODE_BORDER_DIRECTED);
-	emvb->add_margin_child(TTR("Emission Mask"), emission_mask_mode);
+	emission_mask_mode->add_item(TTRC("Solid Pixels"), MASK_MODE_SOLID);
+	emission_mask_mode->add_item(TTRC("Border Pixels"), MASK_MODE_BORDER);
+	emission_mask_mode->connect(SceneStringName(item_selected), callable_mp(this, &Particles2DEditorPlugin::_emission_mask_mode_item_changed));
+	emvb->add_margin_child(TTRC("Mask Mode"), emission_mask_mode);
+
+	emission_direction_mode = memnew(OptionButton);
+	emission_direction_mode->add_item(TTRC("None"), DIRECTION_MODE_NONE);
+	emission_direction_mode->add_item(TTRC("Generate"), DIRECTION_MODE_GENERATE);
+	emission_direction_mode->add_item(TTRC("Texture"), DIRECTION_MODE_TEXTURE);
+	emission_direction_mode->connect(SceneStringName(item_selected), callable_mp(this, &Particles2DEditorPlugin::_validate_textures).unbind(1));
+	emission_direction_mode->set_item_disabled(DIRECTION_MODE_GENERATE, true);
+	emvb->add_margin_child(TTRC("Direction Mode"), emission_direction_mode);
+
+	direction_img_label = memnew(Label);
+	direction_img_label->set_text(TTRC("Direction Texture"));
+	direction_img_label->set_theme_type_variation("HeaderSmall");
+	emvb->add_child(direction_img_label);
+	direction_img_label->hide();
+
+	direction_img_hbox = memnew(HBoxContainer);
+	direction_img_path_line_edit = memnew(LineEdit);
+	direction_img_hbox->add_child(direction_img_path_line_edit);
+	direction_img_path_line_edit->set_h_size_flags(Control::SIZE_EXPAND_FILL);
+	direction_img_path_line_edit->set_editable(false);
+	direction_img_path_line_edit->set_placeholder(vformat(TTRC("Direction texture path")));
```

Comment:
Also should be removed (and why did it use vformat...).

Reply:
c&p :upside_down_face: